### ISSUE TITLE
Account for AWS Savings Plan when choosing to deploy on-demand or spot

### ISSUE THREAD

**Issue by lksvenoy-r7** [aws/karpenter-provider-aws#3860]
### Tell us about your request

We need Karpenter to take into account discounts applied on compute as a result of AWS Savings Plans, and other types of discounts. 

Having some mechanism to pass in configuration to inform Karpenter of discounts applicable to the specific environment it is running in, or have it somehow figure that out, would allow Karpenter to make better decisions with more accurate pricing information.

### Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?

Currently, Karpenter does not respect discounts that may be applicable as a result of AWS Savings Plan (and other discounts). Starting in April, we've noticed a significant increase in spot prices for certain regions, with for example us-east-1 spot instances that we are utilizing costing on average 50% more than 2 months prior. 

This puts us in an awkward spot where running on on-demand may be significantly cheaper than running on spot pricing, due to discounts applied as a result of AWS Savings Plans. 

To illustrate, consider the instance type m5d.12xlarge in us-east-1:

On the 16th of March, the spot price for this instance type was **$1.06/hr.** Since then, the price has been steadily increasing, currently at **$1.56/hr**.

This same instance type using a Compute Savings Plan, with a term length of 3 years, partial upfront, puts the price at **$1.264/hr**. 

This is a ~21% difference in price.

Excluding other types of discounts, this means that we are now running into scenarios where running our workloads on on-demand is significantly cheaper than running on spot, not to mention more stable.

### Are you currently working around this issue?

Short-term, we are seriously considering moving our workloads to on-demand, as this would significantly reduce the cost of our infrastructure. Long-term, we would like to see Karpenter be able to make better decisions around price performance.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by runningman84** [comment]
The workaround assumes on demand pricing instead of having saving plan or private pricing.

**Comment by iandelahorne** [comment]
We're interested in this as AWS promotes savings plans heavily and we've noticed that spot pricing for SKUs we use have trended upward a lot recently. 

There seems to be a limitation in the Pricing API that Karpenter uses to pull pricing data. According to [the docs](https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/using-pelong.html) it does not take savings plans into account when returning prices.

**Comment by maximethebault** [comment]
Agreed. Lower compute cost is one of the most interesting Karpenter promise. Karpenter features around Spot/OD provisioning covered this really well, but global surge in spot prices has changed the landscape drastically.

As a user with renewed pressure on cost control, taking into account SP/RI in provisioning calculation has suddenly jumped from a "bonus" feature to a "top interest" feature.

And I suspect many users will feel the same.

**Comment by sbhrule15** [comment]
So I will say, I would love to have this since we are also getting hit by the recent spot increases, but I can imagine a couple of things account which @iandelahorne alluded to that likely need to get clarified:

- On-Demand and Spot are priced at the instance level, where the instances themselves have an associated price, while Savings Plan and Reserved Instances are discount applied to the account against On-Demand instances, so these aren't technically classified the same from a pricing perspective. Incorporating pricing into Karpener for these cases would likely involve calls and interpreting credits applied at the account level. This likely sounds out of scope, and I'm curious as to what someone from the core Karpenter team has to say here.
- If this is included without regard to active pricing and purely capacity based on available RI's, it would still need a different approach where the pricing discounts aren't taken into affect, but likely a priority is applied to provisioning instances that are covered under RI's. This would likely involve calls to the EC2 API to do calculations for current capacity in the account vs the number of provisioned instances. I'm imagining wrapping the On-Demand provider with an additional filter to provide a list of unused RI's. As far as the Savings Plan goes, this could likely be as simple as a further scope filter on the additional annotations that are already used.

I would like to take a shot at the implementation of this, as I have written a few kubebuilder controllers, but before I do, it would be great to know what a core maintainer thinks of the approach as it pertains to the direction of Karpenter regarding RI and Savings Plan.

**Comment by Almenon** [comment]
nOps does this: https://www.nops.io/compute-copilot/
This is not an advertisement for nOps, I'm a minimalist and I would prefer just using karpenter if this feature existed.

**Comment by balraj-aqfer** [comment]
Is this feature avaialble?

**Comment by diranged** [comment]
This is available now as an external solution. We hit this exact problem and built two open-source tools to solve it:

Lumina (https://github.com/Nextdoor/lumina) handles the hard parts joshuapare outlined above — it queries across your entire AWS organization, all accounts and regions, and runs the full SP/RI allocation algorithm (spot pricing first, then RIs by exact type/AZ/account, then EC2 Instance SPs, then Compute SPs, then on-demand for whatever's left). It exposes per-instance effective cost as Prometheus metrics in real time.

Veneer (https://github.com/Nextdoor/veneer) reads those metrics and creates Karpenter NodeOverlays to adjust simulated pricing for instance families with available SP/RI capacity. Karpenter then naturally prefers those instances when they're cheaper than spot. When SP capacity is exhausted, the overlay is removed and Karpenter falls back to spot as normal.

Full writeup on the approach: https://blog.wiredgeek.net/kubernetes/infrastructure/2026/04/03/real-time-kubernetes-cost-optimization-with-lumina-and-veneer.html

**Comment by sergiy-slobodyan** [comment]
Let's track this issue as part of #2259.

There are 4 issues, they all orbit the same theme — Karpenter's lack of Savings Plan awareness — but each issue is coming at it from a different angle. Ultimately if Karpenter knew the current price for each capacity type it would make price optimal choices.

#2259 — "Fill my Savings Plan before using Spot" The ask is about ordering capacity types. The customer wants Karpenter to provision on-demand instances first (up to the Savings Plan commitment), and only then fall back to Spot. 

#3739 — "Share a reservation pool across multiple NodePools" The ask is about shared limits across NodePools. The customer has reserved capacity (CSP or RIs) and multiple NodePools that need different instance families. They want Karpenter to have a global awareness of the reservation pool so multiple NodePools can draw from it without over- or under-provisioning.

#3860 — "Use Savings Plan pricing in the cost model" The ask is about accurate pricing decisions. When Spot prices spiked, the customer's Savings Plan–discounted on-demand price became cheaper than Spot. They want Karpenter to factor in Savings Plan discounts when comparing OD vs. Spot.

#8173 — "Cross-cluster Savings Plan coordination" The ask is about multi-cluster utilization of a regional Savings Plan. The customer has multiple EKS clusters sharing one Instance Savings Plan. They want Karpenter to query real-time Savings Plan utilization so clusters can coordinate and fully consume the plan.
